levels/lovelace_tower/events.js

```diff
@@ -1,353 +1,162 @@
-    lovelace_key: () => obtainKeySpellComplete("lovelace_spell"),
-    hopper_key: () => obtainKeySpellComplete("hopper_spell"),
-    neumann_key: () => obtainKeySpellComplete("neumann_spell"),
-    turing_key: () => obtainKeySpellComplete("turing_spell"),
-  };
-
-// When the player has retrieved the House Lovelace spell, they are able to enter the Secret Library
-// They are also able to open the library door that leads back to the Main Hall
-// Eventually they will also be able to use the spell to gain access to the next house from the Main Hall
-  const obtainKeySpellComplete = (object) => {
-    unlockObject(object);
-    if (!worldState.insideCatacombs.keySpellsObtained.includes(object))
-      worldState.insideCatacombs.keySpellsObtained.push(object);
-
-    if (worldState.insideCatacombs.keySpellsObtained.length == 4) {
-      world.grantItems(["magic_key"]);
-      unlockObject("magic_key");
+  const worldState = merge(
+    INITIAL_STATE,
+    world.getState(LOVELACE_TOWER_STATE_KEY)
+  );
+
+  // Top object pieces need to be longer so they always depth sort
+  // above the player, even when the player would normally be depth
+  // sorted above them.
+  // Taken from https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=//%20Top%20shelf%20pieces,)%3B
+  world.forEachEntities(
+    ({ instance }) => instance.isTopObject,
+    (shelf) => {
+      shelf.sprite.body.height *= 2;
     }
-  };
-
-  const runObjectiveEffects = ({ objective }) => {
-    objectives[objective]();
-  };
-
-
-// SPELLS
+  );
 
-// This level only requires Lovelace spell animation to be used on two library doors.
-  const spells = {
-    disappear: (event) => disappear(event),
-    move: (event) => move(event),
-    unlock: (event) => unlock(event),
-  };
+  // Taken fom https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=%7D-,//%20Match%20objectives%20to%20shelves%20they%20should%20unlock%20when%20an,%7D,-//%20Hide%20entities
+  const unlockPairs = [
+    ["api-01-local-function", "api-hidden-door-1"],
+    ["api-02-async-await", "api-hidden-door-2"],
+    ["api-03-fetch", "api-hidden-door-3"],
+  ];
 
-  const disappear = (event) => {
-    applyDisappearTween(event.target.group).then(() => {
-      destroyObject(event.target.group);
+  const doorPairs = [["api-04-remote-and-local", "api-door-1"]];
 
-      if (event.target.unlocksObject) unlockObject(event.target.unlocksObject);
-      if (event.target.unlocksTransition)
-        unlockTransition(event.target.unlocksTransition);
-
-      world.stopUsingTool();
-      world.enablePlayerMovement();
+  if (
+    event.name === "objectiveCompleted" ||
+    event.name === "objectiveCompletedAgain"
+  ) {
+    unlockPairs.forEach(([objectiveKey, hiddenDoorKey]) => {
+      if (event.objective === objectiveKey) {
+        worldState.insideLovelaceTower.hiddenEntities.push(hiddenDoorKey);
```

Comment:
This will push duplicates of the hiddenDoorKey each time the objective is completed right? We might want to check if it's already there first:
```js
if (!worldState.insideLovelaceTower.hiddenEntities.includes(hiddenDoorKey)) {

}
```

Reply:
You're absolutely right! I pulled that from [dev fundamentals](https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=%5D%3B-,if%20\(,%7D,-//%20Hide%20entities), so we'll want to update that too when we have the time. I'll make a GitHub Issue for it as a reminder.